src/ext.rs

```diff
@@ -0,0 +1,64 @@
+//! HTTP extensions
+
+use bytes::Bytes;
+#[cfg(feature = "http1")]
+use http::header::{GetAll, HeaderName, IntoHeaderName};
+use http::HeaderMap;
+
+/// A map from header names to their original casing as received in an HTTP message.
+///
+/// If an HTTP/1 response `res` is parsed on a connection whose option
+/// [`http1_preserve_header_case`] was set to true and the response included
+/// the following headers:
+///
+/// ```ignore
+/// x-Bread: Baguette
+/// X-BREAD: Pain
+/// x-bread: Ficelle
+/// ```
+///
+/// Then `res.extensions().get::<HeaderCaseMap>()` will return a map with:
+///
+/// ```ignore
+/// HeaderCaseMap({
+///     "x-bread": ["x-Bread", "X-BREAD", "x-bread"],
+/// })
+/// ```
+///
+/// [`http1_preserve_header_case`]: /client/struct.Client.html#method.http1_preserve_header_case
+#[derive(Clone, Debug, Default)]
+pub struct HeaderCaseMap(HeaderMap<RawHeaderName>);
+
+#[cfg(feature = "http1")]
+impl HeaderCaseMap {
+    /// Returns a view of all spellings associated with that header name,
+    /// in the order they were found.
+    pub fn get_all(&self, name: &HeaderName) -> GetAll<'_, RawHeaderName> {
+        self.0.get_all(name)
+    }
+
+    #[cfg(any(test, feature = "ffi"))]
+    pub(crate) fn insert(&mut self, name: HeaderName, orig: Bytes) {
+        self.0.insert(name, RawHeaderName(orig));
+    }
+
+    pub(crate) fn append<N>(&mut self, name: N, orig: Bytes)
+    where
+        N: IntoHeaderName,
+    {
+        self.0.append(name, RawHeaderName(orig));
+    }
+}
+
+/// Represents a raw header name, which may not in lower case.
+///
+/// Exclusively used by the `HeaderCaseMap` extension.
+#[derive(Clone, Debug)]
+pub struct RawHeaderName(Bytes);
```

Comment:
So before, with a `Bytes` a user could do quite a bit of things with it, not least of all view the actual slice of bytes. I do think there's value in us separating the type some, but probably need an `AsRef<[u8]>` at least.

Do you think returning an `impl AsRef<[u8]>` would be horrible, to start?

Reply:
Nah that's a good idea.

Reply:
Or, I also just had this idea, in case we don't know yet. I thought, what do you want to do with this first. I assume you just want to receive the extension from the server request, and pass it on to the client request, and then same with the client response passed back to the server response. I assume you don't actually need to edit the map really.

If that's so, we could just ditch making this type and `get_all` public, and make that a subject of another PR. :shrug:

Reply:
We could ditch the entire `ext` module even.

Reply:
`ext` and `HeaderCaseMap` are not public items anymore. In my use case we pass requests and responses between client and upstream mostly as is so I don't need anything more.

Reply:
Oh you're right, we don't have to make them public _just yet_, and it still works. Seems like good plan to me. Making them public can be a more detailed proposal :)